frontend/src/js/components/PageViewer/VirtualScroll.tsx

```diff
@@ -45,23 +46,60 @@ export const VirtualScroll: FC<VirtualScrollProps> = ({
   pageNumbersToPreload,
 
   rotation,
+  scale
 }) => {
   // Tweaked this and 2 seems to be a good amount on a regular monitor
   // The fewer pages we preload the faster the initial paint will be
   // Could possibly make it dynamic based on the visible of the container
   const PRELOAD_PAGES = 2;
 
-  const pageHeight = CONTAINER_AND_MARGIN_SIZE;
+  // This must be the same as the margin CSS of .pageContainer
+  const MARGIN = 10;
+
+  const containerSize = 1000 * scale;
+  const pageHeight = containerSize + (MARGIN * 2);
 
   const viewport = useRef<HTMLDivElement>(null);
 
   // TODO: move pageCache up?
-  const [pageCache] = useState(() => new PageCache(uri, searchQuery));
+  const [pageCache] = useState(() => new PageCache(uri, containerSize, searchQuery));
 
   // We have a second tier cache tied to the React component lifecycle for storing
   // rendered pages which allows us to swap out stale pages without flickering pages
   const [renderedPages, setRenderedPages] = useState<RenderedPage[]>([]);
 
+  // denounce is used to avoid multiple re-rendering when user clicks on zoom 
+  // buttons several times and container size changes too quickly 
+  const debouncedRefreshPreview = React.useMemo(
+    () =>
+      debounce((pageCache: PageCache, containerSize: number) => {
+        pageCache.setContainerSize(containerSize);
+
+      setRenderedPages((currentPages) => {
+        const newPages: RenderedPage[] = currentPages.map((page) => {
+          console.log("refreshing page: ", page.pageNumber);
+          const refreshedPage = pageCache.refreshPreview(
+            page.pageNumber,
+            page.getPagePreview,
+            containerSize          
+          ); 
+          return {
+            pageNumber: page.pageNumber,
+            getPagePreview: refreshedPage.preview,
+            getPageData: refreshedPage.data,
+          }
+        });
+
+        return newPages;
+      })
+      }, 500),
+    []
+  );
+
+  useEffect(() => {
+    debouncedRefreshPreview(pageCache, containerSize);
+  }, [pageCache, containerSize, debouncedRefreshPreview]);
+
```

Comment:
This useEffect is used to detect changes on container size. Container size changes when use zooms in or zooms out. Debounce is used to avoid multiple page re-rendering when user clicks zooms in/out multiple times